File: leetcode/05-linkedLists.ts
class ListNode_ {
  val: number
  next: ListNode | null
  constructor(val?: number, next?: ListNode | null) {
    this.val = (val === undefined ? 0 : val)
    this.next = (next === undefined ? null : next)
  }
}

//! Reverses a linked list
/*
	[1,2,3,4,5] -> [5,4,3,2,1]
	[1,2] -> [2,1]
*/

const reverseList = (head: ListNode | null): ListNode | null => {
  let response: ListNode | null = null;
  while (head) {
    const temporal = head.next;
    head.next = response;
    response = head;
    head = temporal
  }
  return response;
}

//! Mergea dos listas enlazadas que estan sorteadas
/*
list1 = [1,2,4], list2 = [1,3,4]
	-> [1,1,2,3,4,4]

list1 = [], list2 = []
	-> []

list1 = [], list2 = [0]
	-> [0]
*/

const mergeTwoLists = (
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null => {
  let response: ListNode = new ListNode(0)
  let currentNode: ListNode = response

  while(list1 && list2) {
    if (list1.val < list2.val) {
      currentNode.next = list1
      list1 = list1.next
    } else {
      currentNode.next = list2;
      list2 = list2.next
    }

    currentNode = currentNode.next
  }

  if(list1) currentNode.next = list1
  if(list2) currentNode.next = list2

  return response.next
}

//! Verifica si una lista enlazada es un palindromo
/*
[1,2,2,1] -> true
[1,2]     -> false
*/

const isPalindrome = (head: ListNode | null): boolean => {
  let
    slow: ListNode | null = head,
    fast: ListNode | null = head,
    reversed: ListNode | null = head;
  
  while(fast && fast.next) {
    fast = fast.next.next

    const temporal: ListNode | null = slow!.next
    slow!.next = reversed
    reversed = slow
    slow = temporal
  }

  if (fast) slow = slow!.next

  while(slow) {
    if(slow.val !== reversed!.val) return false;
    slow = slow.next
    reversed = reversed!.next
  }
  return true
}

//! Remueve un elemento de una lista enlazada
/*
Input: head = [1,2,6,3,4,5,6], val = 6
Output: [1,2,3,4,5]
*/

const removeElements = (head: ListNode | null, val: number): ListNode | null => {
  if(!head) return head
  let current = head

  while(current && current.next)
    if(current.next.val === val) current.next = current.next.next
    else current = current.next

  if(head.val === val) return head.next
  return head
}

//! 83. Remove Duplicates from Sorted List
/*
Input: head = [1,1,2,3,3]
Output: [1,2,3]
*/

const deleteDuplicates = (head: ListNode | null): ListNode | null => {
  if(!head) return null
  let current: ListNode | null = head

  while(current && current.next)
    current.val === current.next.val
      ? current.next = current.next.next
      : current = current.next

  return head
}

//! 143 ReorderList
/*
  Toma una lista y pon el primer elemento al inicio despues el ultimo asi consecutivamente
  Input: head = [1,2,3,4]
  Output: [1,4,2,3]

  Input: head = [1,2,3,4,5]
  Output: [1,5,2,4,3]
*/

function toArray(head: ListNode | null): Array<number> {
  const response: Array<number> = []

  for(let current = head; current; current = current.next)
    response.push(current.val)

  return response
}

function getOrder(list: Array<number>): Array<number> {
  const response: Array<number> = []

  for(let i = 0, j = list.length-1; i <= j; i++, j--) {
    if(i === j) response.push(list[i])

    response.push(list[i])
    response.push(list[j])
  }
  return response
}

function reorderList_(head: ListNode | null) {
  const arrayList: Array<number> = toArray(head)
  const orderedList: Array<number> = getOrder(arrayList)

  for (
    let i = 0, current = head;
    current && i < orderedList.length;
    i++, current = current.next
  )
    current.val = orderedList[i]
}

//? Speed -------------------------------------

function splitLinkedListInHalf(list: ListNodeNull): Array<ListNodeNull> {
  let slow = list, fast = list!.next
  while(fast && fast.next) {
    slow = slow!.next
    fast = fast.next.next
  }
  const secondHalf = slow!.next
  slow!.next = null

  return [list, secondHalf]
}

function reverse_(list: ListNodeNull): ListNodeNull {
  let response: ListNodeNull = null
  
  while(list) {
    const temporal = list.next
    list.next = response
    response = list
    list = temporal
  }

  return response
}

function mergeLists(list1: ListNodeNull, list2: ListNodeNull): ListNodeNull {
  let response: ListNodeNull = new ListNode()
  let pointer = response

  while(list1 && list2) {
    pointer.next = list1
    list1 = list1.next
    pointer = pointer.next
    pointer.next = list2
    list2 = list2.next
    pointer = pointer.next
  }

  if(list1) pointer.next = list1
  if(list2) pointer.next = list2
  return response.next
}

function reorderList(list: ListNodeNull): ListNodeNull {
  let [firstHalf, secondHalf] = splitLinkedListInHalf(list)
  let reversedSecondHalf = reverse(secondHalf)
  return mergeLists(firstHalf, reversedSecondHalf)
}

//! 160. Intersection of Two Linked Lists
/*
l1 = [4,1,8,4,5], l2 = [5,6,1,8,4,5]
output [8,4,5]
*/

function getIntersectionNode(list1: ListNodeNull, list2: ListNodeNull): ListNodeNull {
  let
    pointer1: ListNodeNull = list1,
    pointer2: ListNodeNull = list2
  while(pointer1 != pointer2) {
    pointer1 = pointer1 ? pointer1.next : list2
    pointer2 = pointer2 ? pointer2.next : list1
  }
  return pointer1
}

//! 2130. Maximum Twin Sum of a Linked List
/*
Input: head = [5,4,2,1]
Output: 6
Explanation:
Nodes 0 and 1 are the twins of nodes 3 and 2, respectively. All have twin sum = 6.
There are no other nodes with twins in the linked list.
Thus, the maximum twin sum of the linked list is 6.

Input: head = [4,2,2,3]
Output: 7
Explanation:
The nodes with twins present in this linked list are:
- Node 0 is the twin of node 3 having a twin sum of 4 + 3 = 7.
- Node 1 is the twin of node 2 having a twin sum of 2 + 2 = 4.
Thus, the maximum twin sum of the linked list is max(7, 4) = 7.
*/

function pairSum(head: ListNode | null): number {
  let
    slow: ListNode | null = head,
    fast: ListNode | null = head,
    reversed: ListNode | null = null;

  while(fast && fast.next) {
    fast = fast.next.next
    const temporal = slow!.next
    slow!.next = reversed
    reversed = slow
    slow = temporal
  }

  let response = 0

  while(slow) {
    const currentValue = reversed!.val + slow!.val
    if (currentValue > response) response = currentValue
    reversed = reversed!.next
    slow = slow!.next
  }
  return response
}

//! 19 Remove Nth Node From End of List
/*
Input: head = [1,2,3,4,5], n = 2
Output: [1,2,3,5]

Input: head = [1,2], n = 1
Output: [1]
*/

function removeNthFromEnd(head: ListNode | null, n: number): ListNode | null {
  let response: ListNode | null = new ListNode(0, head),
    left: ListNode | null = response,
    right: ListNode | null = head;

    while(n > 0) {
      right = right!.next
      n--
    }

    while(left && right) {
      left = left.next
      right = right.next
    }

    left!.next = left!.next!.next
    return response.next
}

//! 1721 Swapping Nodes in a Linked List
/*
Input: head = [1,2,3,4,5], k = 2
Output: [1,4,3,2,5]

Input: head = [7,9,6,6,7,8,3,0,9,5], k = 5
Output: [7,9,6,6,8,7,3,0,9,5]
*/

function swapNodes(head: ListNode | null, k: number): ListNode | null {
  let dummy = new ListNode(0, head),
    left: ListNode | null = dummy,
    right: ListNode | null = dummy,
    final: ListNode | null = dummy

  for(let i = 0; i < k; i++) {
    left = left!.next
  }

  while(k > 0) {
    final = final!.next
    k--
  }

  while(right && final) {
    right = right.next
    final = final.next
  }

  // Swap values
  const temporal = left!.val
  left!.val = right!.val
  right!.val = temporal

  return dummy.next
}

//! 138 Copy List with random pointer
/*
tienes una lista enlazada que contiene un valor y un valor random haz una copia de ella

Input: head = [[7,null],[13,0],[11,4],[10,2],[1,0]]
Output: [[7,null],[13,0],[11,4],[10,2],[1,0]]

Input: head = [[1,1],[2,1]]
Output: [[1,1],[2,1]]

Input: head = [[3,null],[3,0],[3,null]]
Output: [[3,null],[3,0],[3,null]]
*/

//? review

class NodeRandom {
  val: number
  next: NodeRandom | null
  random: NodeRandom | null
  constructor(val?: number, next?: NodeRandom, random?: NodeRandom) {
    this.val = val ? val : 0
    this.next = next ? next : null
    this.random = random ? random : null
  }
}

function copyRandomList(head: NodeRandom | null): NodeRandom | null {
  let hash = new Map<NodeRandom, NodeRandom>();

  let response = new NodeRandom(0);
  let current = head;
  let newCurrent = response;

  while (current) {
      let newNode = new NodeRandom(current.val);
      newCurrent.next = newNode;
      newCurrent = newNode;
      hash.set(current, newNode);

      current = current.next;
  }

  current = head;
  while (current) {
      let inNew = hash.get(current);
      if (current.random) {
          let inNewRand = hash.get(current.random);
          inNew!.random = inNewRand;
      }
      current = current!.next;
  }

  return response.next;
}

const copyRandomListd = (head: NodeRandom | null): NodeRandom | null => {
  let hash = new Map<NodeRandom, NodeRandom>()

  let response = new NodeRandom(0)
  let current = head
  let newCurrent = response

  while(current) {
    let newNode = new NodeRandom(current.val)
    newCurrent.next = newNode
    newCurrent = newNode

    current = current.next
  }

  current = head
  while(current) {
    let inNew = hash.get(current)
    if(current.random) {
      let inNewRand = hash.get(current.random)
      inNew!.random = inNewRand
    }
    current = current!.next
  }
  return response.next
}

//! 707 Design Linked list

interface Node {
  val: number
  next: Node
}

class MyLinkedList {
  head: Node;
  size: number;
  
  constructor() {
    this.head = null
    this.size = 0
  }

  getNode(index: number): Node | null {
    if (index < 0 || index >= this.size) return null
    let count = 0
    let node = this.head
    while (count < index) {
      node = node.next
      count++
    }
    return node
  }
  
  get(index: number): number {
    const node = this.getNode(index)
    if (!node) return -1
    return node.val
  }

  addAtHead(val: number): void {
    const next = this.head
    this.head = { val, next }
    this.size++
  }

  addAtTail(val: number): void {
    if (this.size === 0) {
      this.addAtHead(val)
      return
    }
    const node = this.getNode(this.size - 1)
    node.next = { val, next: null }
    this.size++
  }

  addAtIndex(index: number, val: number): void {
    if (index < 0 || index > this.size) return
    if (index === this.size) {
      this.addAtTail(val)
      return
    }
    if (index === 0) {
      this.addAtHead(val)
      return
    }
    const node = this.getNode(index - 1)
    node.next = { val, next: node.next}
    this.size++
  }

  deleteAtIndex(index: number): void {
    if (index < 0 || index >= this.size) return
    if (this.size === 0) return
    if (index === 0) {
      this.head = this.head.next
      this.size--
      return
    }
    const node = this.getNode(index - 1)
    node.next = node.next?.next || null
    this.size--
  }
}

//! 2 add two numbers

/*
Input: l1 = [2,4,3], l2 = [5,6,4]
Output: [7,0,8]
Explanation: 342 + 465 = 807.

Input: l1 = [0], l2 = [0]
Output: [0]

Input: l1 = [9,9,9,9,9,9,9], l2 = [9,9,9,9]
Output: [8,9,9,9,0,0,0,1]
*/

function addTwoNumbers(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  let
    dummyNode: ListNode = new ListNode(),
    currentNode: ListNode = dummyNode,
    carry: number = 0;

  while (list1 || list2 || carry !== 0) {
      const
        value1 = list1 ? list1.val : 0,
        value2 = list2 ? list2.val : 0,
        sum: number = value1 + value2 + carry,
        value: number = sum % 10,
        newLinkedList: ListNode = new ListNode(value);

      currentNode.next = newLinkedList;
      currentNode = newLinkedList;
      carry = Math.floor(sum / 10);

      list1 = list1 ? list1.next : null;
      list2 = list2 ? list2.next : null;
  }
  return dummyNode.next;
}

//! 24 Swwap Nodes in Pairs
/*
Input: head = [1,2,3,4]
Output: [2,1,4,3]
Example 2:

Input: head = []
Output: []
Example 3:

Input: head = [1]
Output: [1]
*/

function swapPairs(head: ListNode | null): ListNode | null {
  let dummy = new ListNode(0, head)
  let previous = dummy, current = head

  while(current && current.next) {
    // Saving pointers
    const nextNext: ListNode | null = current.next.next
    const next: ListNode | null = current.next

    // Reverse pointers
    next.next = current
    current.next = nextNext
    previous.next = next

    // Update pointers
    previous = current
    current = nextNext
  }
  return dummy.next
}

File: leetcode/05-linkedLists.go
package main
func main() { }

type ListNode struct {
	Val int
	Next *ListNode
}

//! Reverses a linked list
/*
	[1,2,3,4,5] -> [5,4,3,2,1]
	[1,2] -> [2,1]
*/


func reverseList(head *ListNode) *ListNode {
	var response *ListNode
	for current := head; current != nil; current = current.Next {
		response = &ListNode{Val: current.Val, Next: response}
	}
	return response
}

//? Type
func reverseList_m(head *ListNode) *ListNode {
	var response *ListNode
	curr := head

	for curr != nil {
		tmp := curr.Next
		curr.Next = response
		response = curr
		curr = tmp
	}

	return response
}

//! Mergea dos listas enlazadas que estan sorteadas
/*
list1 = [1,2,4], list2 = [1,3,4]
	-> [1,1,2,3,4,4]

list1 = [], list2 = []
	-> []

list1 = [], list2 = [0]
	-> [0]
*/

func mergeTwoLists(list1, list2 *ListNode) *ListNode {
	var response *ListNode
	current1 := list1
	current2 := list2

	for current1 != nil && current2 != nil {
		if current1.Val > current2.Val {
			response = &ListNode{Val: current2.Val, Next: response}
			current2 = current2.Next
		} else {
			response = &ListNode{Val: current1.Val, Next: response}
			current1 = current1.Next
		}
	}

	for current1 != nil {
		response = &ListNode{Val: current1.Val, Next: response}
		current1 = current1.Next
	}
	for current2 != nil {
		response = &ListNode{Val: current2.Val, Next: response}
		current2 = current2.Next
	}
	return reverseList(response)
}

//? Speed
func mergeTwoLists_s(list1, list2 *ListNode) *ListNode {
	dummy := &ListNode{}
	node := dummy

	for list1 != nil && list2 != nil {
		if list1.Val < list2.Val {
			node.Next = list1
			list1 = list1.Next
		} else {
			node.Next = list2
			list2 = list2.Next
		}
		node = node.Next
	}

	if list1 != nil {
		node.Next = list1
	} else if list2 != nil {
		node.Next = list2
	}

	return dummy.Next
}

//! Verifica si una lista enlazada es un palindromo
/*
[1,2,2,1] -> true
[1,2]     -> false
*/

func isPalindrome_b(head *ListNode) bool {
	head2 := reverseList(head)
	for head2 != nil && head != nil {
		if head.Val != head2.Val { return false }
		head, head2 = head.Next, head2.Next
	}
	return true
}

func isPalindrome(head *ListNode) bool {
	slow, fast := head, head
	var reversed *ListNode
	
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		temporal := slow.Next
		slow.Next = reversed
		reversed = slow
		slow = temporal
	}

	if fast != nil { slow = slow.Next }

	for slow != nil {
		if slow.Val != reversed.Val { return false }
		slow, reversed = slow.Next, reversed.Next
	}

	return true
}

//! Remueve un elemento de una lista enlazada
/*
Input: head = [1,2,6,3,4,5,6], val = 6
Output: [1,2,3,4,5]
*/

func removeElements(head *ListNode, val int) *ListNode {
	if head == nil { return nil }

	for curr := head; curr.Next != nil; {
		if curr.Next.Val == val {
			curr.Next = curr.Next.Next
		} else {
			curr = curr.Next
		}
	}

	if head.Val == val {
		return head.Next
	}

	return head
}

//! 83. Remove Duplicates from Sorted List
/*
Input: head = [1,1,2,3,3]
Output: [1,2,3]
*/

func deleteDuplicates(head *ListNode) *ListNode {
	if head == nil { return nil }
	for current := head; current.Next != nil; {
		if current.Val == current.Next.Val {
			current.Next = current.Next.Next
		} else {
			current = current.Next
		}
	}
	return head
}

//! Retorna el punto medio de una lista
/*
	[1,2,3,4,5] ->   [3,4,5]
	[1,2,3,4,5,6] -> [4,5,6]
*/

func middleNode(head *ListNode) *ListNode {
	slow, fast := head, head
	for slow != nil && fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	return slow
}

//! 143 ReorderList
/*
  Toma una lista y pon el primer elemento al inicio despues el ultimo asi consecutivamente
  Input: head = [1,2,3,4]
  Output: [1,4,2,3]

  Input: head = [1,2,3,4,5]
  Output: [1,5,2,4,3]
*/
func toArray(head *ListNode) []int {
	var response []int
	for current := head; current != nil; current = current.Next {
		response = append(response, current.Val)
	}
	return response
}

func getOrder(list []int) []int {
	var response []int
	for i, j := 0, len(list)-1; i <= j; i, j = i+1, j-1 {
		if i == j {
			response = append(response, list[i])
			break
		}
		response = append(response, list[i])
		response = append(response, list[j])
	}
	return response
}

func reorderList(head *ListNode) {
	arrayList := toArray(head)
	orderedList := getOrder(arrayList)
	//fmt.Println(orderedList)

	for i, current := 0, head;
	current != nil && i < len(orderedList);
	i, current = i+1, current.Next {
		current.Val = orderedList[i]
	}
}